dmutils/email.py

```diff
@@ -231,3 +232,95 @@ def decode_invitation_token(encoded_token):
 
 def token_created_before_password_last_changed(token_timestamp, user_timestamp):
     return token_timestamp < user_timestamp
+
+
+class DMNotifyClient(object):
+    """Digital Marketplace wrapper around the Notify python client."""
+
+    client_class = NotificationsAPIClient
+
+    def __init__(self, api_key, base_url='https://api.notifications.service.gov.uk', logger=None):
+        """Set up logging and mail client."""
+        self.logger = logger or current_app.logger
+        self.client = self.client_class(api_key, base_url)
+
+    @staticmethod
+    def get_reference(email_address, template_id, personalisation, time=None, set_time=True, **kwargs):
+        """
+        Method to return the standard reference given the variables the email is sent with.
+
+        :param email_address: Emails recipient
+        :param template_id: Emails template ID on Notify
+        :param personalisation: Template parameters
+        :param time: Generally not wise to set this. Spoofs time that the email was sent.
+        :param set_time: Flag, whether or not to set the time in the reference.
+        :param kwargs: Extra data passed to the reference eg. {'notes': 'Manual resend'}
+        :return: String json 'reference' to be passed to client.send_email_notification or self.send_email
+        """
+        if time is None and set_time:
+            time = datetime.now().strftime(DATETIME_FORMAT)
+        details_dict = OrderedDict([
+            ('email_address', email_address),
+            ('template_id', template_id),
+            ('personalisation', personalisation),
+            ('time', time),
+        ])
+        details_dict.update([(k, v) for k, v in kwargs.items()])
+        return json.dumps(details_dict)
```

Comment:
So the problem with this is that we have no way of getting the `personalisation` context dict back from the notify api unless we push it into the reference. Therefore unless we store it we can't know what context the email was sent with :/

The `email_address`, `template_id` and `personalisation` combined denote a unique instance of an email.
We may want to send the same user the same template with 2 different contexts but we have no way of knowing which is which unless we store the personalisation in the reference.

For instance these are 2 different emails, with potentially vastly different content:
```
send_email('example@example.com', '1234-12341234-12341234-1234', {'foo': 'bar'})
send_email('example@example.com', '1234-12341234-12341234-1234', {'foo': 'baz'})
```
But we can't tell which is which without the personalisation.

#### Could probably be considered a premature optimisation